pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -326,6 +339,11 @@ public SegmentMetadata getSegmentMetadata(String tableNameWithType, String segme
     }
   }
 
+  @Override
+  public Set<String> getColumnNamesByTable(String tableNameWithType) {
+    return _tableColumnNamesMap.computeIfAbsent(tableNameWithType, k -> Collections.emptySet());
```

Comment:
Can this case memory issue if there are too many tables on an instance? We can check how the overhead of computing the map (if we can do once per query, not segment) compares to the overhead of caching (especially for large MT cluster).

Also, what happens when a table is deleted, is this entry cleared?

Reply:
I don't quite fully understand the meaning of testing the overhead of computing the map once per query. The table columns are fetched from the schema which is stored in Zookeeper, and they are used to detect the column name mismatch during query execution. If schema fetch is done once per query, the ZK access would be a bottleneck for query execution. Plus, it'd be good to stick to one behavior in all different circumstances. Having a cache for high qps usecase fabric while directly fetching once per query for large MT cluster would make the code more difficult to maintain.

`_tableColumnNamesMap` would stay the same state with `_tableDataManagerMap`. This is to make sure that the behavior of these two object are always consistent. And I don't see `_tableDataManagerMap` drops any reference of dataManager, until the server is shutting down.